packages/typespec-test/test/unionwidget_dpg/generated/typespec-ts/src/utils/deserializeUtil.ts

```diff
@@ -0,0 +1,512 @@
+/** type predict function fpr WidgetData5Output array from WidgetData0Output | WidgetData5Output[] */
+function isWidgetData5Array(
+  obj: WidgetData0Output | WidgetData5Output[]
+): obj is WidgetData5Output[];
+/** type predict function fpr WidgetData5Output array from WidgetData0Output[] | WidgetData5Output[] */
+function isWidgetData5Array(
+  obj:
+    | WidgetData0Output[]
+    | WidgetData5Output[]
+    | WidgetData0Output
+    | WidgetData5Output[]
+): obj is WidgetData5Output[] {
+  if (Array.isArray(obj) && obj.length > 0) {
+    return (obj as WidgetData5Output[])[0].data !== undefined;
+  }
+
+  return false;
+}
+
+/** deserialize function for WidgetData5 array */
+function deserializeWidgetData5Array(obj: WidgetData5Output[]): WidgetData5[] {
+  return (obj || []).map((item) => {
+    return {
+      data:
+        typeof item["data"] === "string"
+          ? stringToUint8Array(item["data"], "base64")
+          : item["data"],
+    };
+  });
+}
+
+/** deserialize function for WidgetData0Output[] | WidgetData5Output[] */
+export function deserializeWidgetData0ArrayAndWidgetData5ArrayUnion(
+  obj: WidgetData0Output[] | WidgetData5Output[]
+): WidgetData0[] | WidgetData5[] {
+  if (isWidgetData5Array(obj)) {
+    return deserializeWidgetData5Array(obj);
+  }
+  return obj;
+}
+
+/** deserialize function for WidgetData0Output[] | WidgetData2Output */
+export function deserializeWidgetData0ArrayAndWidgetData2Union(
+  obj: WidgetData0Output[] | WidgetData2Output
+): WidgetData0[] | WidgetData2 {
+  if (isWidgetData2(obj)) {
+    return deserializeWidgetData2(obj);
+  }
+  return obj;
+}
+
+/** deserialize function for WidgetData0Output | WidgetData2Output[] */
+export function deserializeWidgetData0AndWidgetData2ArrayUnion(
+  obj: WidgetData0Output | WidgetData2Output[]
+): WidgetData0 | WidgetData2[] {
+  if (isWidgetData2Array(obj)) {
+    return deserializeWidgetData2Array(obj);
+  }
+  return obj;
+}
+
+/** deserialize function for WidgetData1Output[] | WidgetData2Output */
+export function deserializeWidgetData1ArrayAndWidgetData2Union(
+  obj: WidgetData1Output[] | WidgetData2Output
+): WidgetData1[] | WidgetData2 {
+  if (isWidgetData2(obj)) {
+    return deserializeWidgetData2(obj);
+  }
+  return obj;
+}
+
+/** deserialize function for WidgetData1Output | WidgetData2Output[] */
+export function deserializeWidgetData1AndWidgetData2ArrayUnion(
+  obj: WidgetData1Output | WidgetData2Output[]
+): WidgetData1 | WidgetData2[] {
+  if (isWidgetData2Array(obj)) {
+    return deserializeWidgetData2Array(obj);
+  }
+  return obj;
+}
+
+/** deserialize function for WidgetData2Output[] | WidgetData3Output */
+export function deserializeWidgetData2ArrayAndWidgetData3Union(
+  obj: WidgetData2Output[] | WidgetData3Output
+): WidgetData2[] | WidgetData3 {
+  if (isWidgetData2Array(obj)) {
+    return deserializeWidgetData2Array(obj);
+  }
+  if (isWidgetData3(obj)) {
+    return deserializeWidgetData3(obj);
+  }
+  return obj;
```

Comment:
I notice that in `deserializeWidgetData2ArrayAndWidgetData3Union` we have two `if`s, but in `deserializeWidgetData1AndWidgetData2ArrayUnion` we only one `if`.

Reply:
Because WidgetData2 is defined as  
```
model WidgetData2 {
  @projectedName("json", "foo_prop")
  fooProp: string;
}
```
and WidgetData3 is defined as 
```
model WidgetData3 {
  @projectedName("json", "bar_prop")
  barProp: string;
}
```
both of them belong to special union variant, each type needs a special deserialize function. 
Whereas WidgetData1 is defined as 
```
model WidgetData1 {
  barProp: string;
}
```
which is just a normal object we don't need special deserialization for it. 
How many if statements depends on how many special union variants we have in this union, not by the number of it.